lib/OpenQA/WebAPI/Controller/Test.pm

```diff
@@ -223,8 +223,55 @@ sub _stash_job_and_module_list {
 sub details {
     my ($self) = @_;
 
-    my $job = $self->_stash_job_and_module_list or return $self->reply->not_found;
-    $self->render($job->should_show_autoinst_log ? 'test/autoinst_log_within_details' : 'test/details');
+    return $self->reply->not_found unless my $job = $self->_stash_job;
+
+    if ($job->should_show_autoinst_log) {
+        my $log = $self->render_to_string('test/autoinst_log_within_details');
+        return $self->render(json => {snippets => {header => $log}});
+    }
+
+    my $modules = read_test_modules($job);
+    my @ret;
+
+    for my $module (@{$modules->{modules}}) {
+        for my $step (@{$module->{details}}) {
+            delete $step->{needles};
+        }
+
+        my $hash = {
+            name           => $module->{name},
+            category       => $module->{category},
+            result         => $module->{result},
+            execution_time => $module->{execution_time},
+            details        => $module->{details},
+            flags          => []};
+
+        for my $flag (qw(important fatal milestone always_rollback)) {
+            if ($module->{$flag}) {
+                push(@{$hash->{flags}}, $flag);
+            }
+        }
+
+        push @ret, $hash;
+    }
+
+    my %tplargs     = (moduleid => '$MODULE$', stepid => '$STEP$');
+    my $brand       = $self->app->config->{global}->{branding};
+    my $bug_actions = $self->render_to_string("branding/$brand/external_reporting", %tplargs);
+
+    if (!$bug_actions) {
+        $bug_actions = $self->render_to_string("branding/plain/external_reporting", %tplargs);
+    }
+
+    my $snips = {
+        header        => $self->render_to_string('test/details'),
+        bug_actions   => $bug_actions,
+        src_url       => $self->url_for('src_step', testid => $job->id, moduleid => '$MODULE$', stepid => 1),
+        module_url    => $self->url_for('step', testid => $job->id, %tplargs),
+        md5thumb_url  => $self->url_for('thumb_image', md5_dirname => '$DIRNAME$', md5_basename => '$BASENAME$'),
+        thumbnail_url => $self->url_for('test_thumbnail', testid => $job->id, filename => '$FILENAME$')};
+
+    return $self->render(json => {snippets => $snips, modules => \@ret});
```

Comment:
Regarding the whole section I wonder if we can keep the former concise definition consisting of just three lines instead of this rather big block now but I do not have specific suggestion right now

Reply:
Not unless you want to keep a duplicate URL mapping in JavaScript.